zebra-chain/src/serialization/read_zcash.rs

```diff
@@ -162,3 +162,20 @@ pub fn canonical_ip_addr(v6_addr: &Ipv6Addr) -> IpAddr {
         Some(_) | None => V6(*v6_addr),
     }
 }
+
+/// Transform a `SocketAddr` into a canonical Zebra `SocketAddr`, converting
+/// IPv6-mapped IPv4 addresses, and removing IPv6 scope IDs and flow information.
+///
+/// See [`canonical_ip_addr`] for detailed info on IPv6-mapped IPv4 addresses.
+pub fn canonical_socket_addr(socket_addr: impl Into<SocketAddr>) -> SocketAddr {
+    use SocketAddr::*;
+
+    let mut socket_addr = socket_addr.into();
+    if let V6(v6_socket_addr) = socket_addr {
+        let canonical_ip = canonical_ip_addr(v6_socket_addr.ip());
+        // creating a new SocketAddr removes scope IDs and flow information
+        socket_addr = SocketAddr::new(canonical_ip, socket_addr.port());
```

Comment:
Some of our proptests were failing, because sanitization removes the scope ID and flow info fields on `SocketAddrIPv6`.

Since we use the `SocketAddr` as a unique `AddressBook` index, we need all addresses in the address book to be canonical. (Otherwise we can have multiple address book entries for the same peer, and connect to it multiple times.)